src/audio/pipewire.c

```diff
@@ -0,0 +1,419 @@
+    // this will be done using the following two step algorythm:
+    // we first get the read index of the ringbuffer, which also returns the amount available in there
+    // if we have enough data in the ringbuffer, aka if the read index isn't less than requested, then we load all of it at once in pipewire
+    // however, if the remainder between the filled part of the buffer and the value of requested is greatter than 0, then we fill that part of the buffer with silence, while taking everything else, hoping we will have more samples next time
+    available_bytes_in_speech_dispatcher_buffer = spa_ringbuffer_get_read_index(&state->rb, &read_index);
+    message(4, "we have %i bytes in the spd ringbuffer", available_bytes_in_speech_dispatcher_buffer);
+    total_number_of_bytes_in_pipewire_buffer = buf->datas[0].maxsize; // how many samples, in bytes,  of audio are in this chunk. Technically, requested itself could be used, but this is done instead to avoid some very rare, but important, edge cases, for example buggy hardware drivers which don't probe the card correctly, so on_process gets called very often, with a value of requested samples greatter than what pipewire could allocate, in maxsize
+// if our pipewire is new enough, we know exactly how many samples to push, given through the value of the requested field, from the structure returned when we dequeue a buffer for filling
+#if PW_CHECK_VERSION(0, 3, 49)
+    if (b->requested > 0)
+    {
+        total_number_of_bytes_in_pipewire_buffer = SPA_MIN(total_number_of_bytes_in_pipewire_buffer, b->requested * state->stride); // because this value is given in frames/samples, and we want bytes
+    }
+// otherwise however, we have no idea how much we should push, so we push as much as we have in the ring buffer, making sure we don't push more than maxsize. If that ends up being less than what would have been requested if we had access to that, we have no choice but to underrun
+#else
+    total_number_of_bytes_in_pipewire_buffer = SPA_MIN(available_bytes_in_speech_dispatcher_buffer, buf->datas[0].maxsize * state->stride); // bytes, not frames)
+#endif
+    message(4, "we should fill %i bytes in the pipewire provided buffer", total_number_of_bytes_in_pipewire_buffer)
+        // if there's something to be read from the ring buffer at this time, we do so now
+        if (available_bytes_in_speech_dispatcher_buffer > 0)
+    {
+        available_for_loading = SPA_MIN(available_bytes_in_speech_dispatcher_buffer, total_number_of_bytes_in_pipewire_buffer);
+        message(4, "we can load %i bytes for now", available_for_loading);
+    }
+    else
+    {
+        available_for_loading = 0;
+        message(4, "nothing to be loaded");
+    }
+    // if there's anything else to fill and the ring buffer doesn't contain enough at this point, fill the difference with silence
+    must_load_with_silence = total_number_of_bytes_in_pipewire_buffer - available_for_loading;
```

Comment:
You can move this computation inside `if (b->requested > 0)`

Reply:
so, the must_load_with_silence variable? or the whole block, where the dependencies of that variable are calculated?

Reply:
I'm only talking about the computation of must_load_with_silence.
available_for_loading and total_number_of_bytes_in_pipewire_buffer are used below so their computation need to be left as such.

Reply:
will do...actually, done!